keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,823 @@
+
+### Evaluation
+
+The job-controller evaluates the successPolicies follow the below rules in order every when pod conditions are updated:
+
+- `succeededIndexes`: the job-controller evaluates `.status.completedIndexes` to see if a set of indexes is there.
+- `succeededCount`: the job-controller evaluates `.status.succeeded` to see if the value is `succeededCount` or more.
+
+After that, the job-controller adds a `SuccessCriteriaMet` condition instead of a `Failed` condition to `.status.conditions`
+and the job-controller terminates still running pods. At that time, if the `successCriterionMetReasonSuffix` is specified,
+the job-controller adds the `JobSuccessPolicy` and the specified suffix combined with `As` to the `reason` field.
+In others, only `JobSuccessPolicy` is added to the `reason` field.
+
+Note that when the job meets one of successPolicies, other successPolicies are ignored.
+
+Finally, the job-controller adds a `Complete` condition to `.status.conditions`.
+
+If any successPolicy isn't set, the job-controller adds both a `SuccessCriteriaMet` and a `Complete` conditions to the Job
+after the job finished.
+
+### Transition of "status.conditions"
+
+When the job with successPolicies is submitted, the job `status.conditions` transits in the following:
+Note that the Job doesn't have an actual `Running` condition in the `status.conditions`.
+
+```mermaid
+stateDiagram-v2
+    [*] --> Running
+    Running --> Failed: Exceeded backoffLimit\n or Matched FailurePolicy
+    Failed --> [*]
+    Running --> SuccessCriteriaMet: Matched SuccessPolicy
+    SuccessCriteriaMet --> Complete: All pods were terminated by successPolicies
+    Complete --> [*]
+```
+
+It means that the job's `.status.conditions` follows the following rules:  
+
+- The job could have both `SuccessCriteriaMet=true` and `Complete=true` conditions.
+- The job couldn't have both `Failed=true` and `SuccessCriteriaMet=true` conditions.
+- The job couldn't have both `Failed=true` and `Complete=true` conditions.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with other terminating policies such as the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+, backoffLimit, and [backoffLimitPerIndex](https://kubernetes.io/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
+in the following situations:
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+- when the job meets the successPolicy and the number of failed pods per indexes exceeds `backoffLimitPerIndex` in all indexes.
+
+To avoid the above conflicts, terminating policies are evaluated the first before successPolicies.
+This means that the terminating policies are respected rather than the successPolicies.
```

Comment:
what if we already added `SuccessCriteriaMet=true`?
Would we still evaluate failure policies?

Reply:
Let me give an example:
I have a success policy `- succeededIndexes: 0`, and a failure policy matching exit code 1.
Index 0 terminates successfully. So at this point we add `SuccessCriteriaMet=true`, and delete the remaining running pods.
Imagine that one of these pods terminates with exit code 1.
What happens? Will the Job be declared Completed or Failed?

Reply:
@alculquicondor That's a good question.
Actually, I talked about the similar situations with @mimowo and @deads2k.
As a result, we agreed that the status (successCriteriaMet vs Failed) must never switch.

Because switching status would bring confusion to systems, which depends on Job API.

Reply:
ok, so once `SuccessCriteriaMet=true`, the Job will end with `Complete=true`, even if later pods could potentially match the failure policy.

That is ok with me. But the KEP contents don't make it very clear.

Reply:
> That is ok with me. But the KEP contents don't make it very clear.

It makes sense. I will add it to this KEP.

Reply:
Done.